Culmination of two years of fixes, especially for export

hybridherbst opened the pull request:
The UnityGLTF repo has been stale for around 2 years now. 
This is approximately the same time when we (prefrontal cortex) started working with glTF extensively, and thus we forked this repository and starting to improve things. There have been more than 300 commits since then, numerous features have been added, and most notably, the package has been used in many productions and battle-tested, especially for _exporting_ glTF files from Unity. Early on, we tried to PR our changes in small chunks, but it soon became clear that the repo was abandoned. 
This PR moves all of our changes over to the main repository and effectively promotes our fork, so that veryone can benefit from the changes done so far. 

## State of glTF in Unity

UnityGLTF hasn't received official support since early 2020. However, a number of forks have fixed issues and improved several key areas, especially animation support,export workflows, color spaces and extendibility. These forks have now been merged back into main so that everyone can benefit from then, and to enable further work.

A separate glTF implementation for Unity, [glTFast](https://github.com/atteneder/glTFast), is on its path towards becoming feature complete, with import already being complete. It leverages modern Unity features such as Burst and Jobs, has better compression support (importing compressed textures and meshes), and also has wider Render Pipeline support, notably supporting URP and HDRP import (and partial export).

glTFast and UnityGLTF can coexist in the same project; you can for example use glTFast for import and UnityGLTF for export, where each of these shine. In fact, we've been contributors to both projects.

- UnityGLTF has very good export support (runtime, editor, animations).
- glTFast has better general import support (compression extensions supported, SRP support).
- If you're playing with custom extensions, UnityGLTF might still be the right choice for import.

## Notable changes

- editor animation export support, originally based on a Sketchfab fork but much improved
- runtime animation export through Timeline Recorder Tracks and also via API
- proper colorspace export and import, for colors, vertex colors, textures, emissive, ...
- export of arbitrary textures, readable or not, including RenderTextures
- exporting blend shapes and blend shape animations with sparse accessors
- numerous hooks for custom extension export (still work in progress – the goal is that what's needed for custom extensions is made public)
- support for KHR_lights_punctual, KHR_materials_unlit
- restructured as UPM package (proper repository restructuring and modernization can happen after this PR)
- GLTFSerialization is now an AsmDef, not a separate sln anymore (easier cross-platform development)
- removed embedded Newtonsoft.Json, using the one from Unity's registry now for better compatibility with other packages
- ability for in-memory export of GLB files (relevant for WebGL and other platforms with limited filesystem access)

## Notable things that haven't changed

- still not fully spec-compliant (see [Known Issues](https://github.com/prefrontalcortex/UnityGLTF#known-issues))
- mixed bag of features that are supported on export and import
- no URP/HDRP import support – use glTFast if you need that
- no support for Draco, Meshopt, KTX2 import or export – use glTFast if you need that

## How to test this PR

1. Open `Window > Package Manager`
2. In Package Manager, click <kbd>+</kbd> and select <kbd>Add Package from git URL</kbd>
3. Paste the following link to the dev branch of this repository:
  ```
  https://github.com/KhronosGroup/UnityGLTF.git?path=/UnityGLTF/Assets/UnityGLTF#dev
  ```
4. Click <kbd>Add</kbd>.

Once the PR is approved, the package will be updated on [OpenUPM](https://openupm.com/packages/org.khronos.unitygltf) as well for easier consumption (no git installation required).

## Example exports

All these projects have been exported with our fork of UnityGLTF:
- https://sketchfab.com/3d-models/biolum-meet-the-giant-isopod-b5cbe237dd554396bf99bdff44601060
- https://gruen.deutschebahn.com/de/dialog/apps/umweltrallye/station2
- https://prefrontalcortex.de/projects/siemensstadt/
- https://bike.needle.tools/
- https://songsofcultures.com/en/ar-collection

and many more.

## Next Steps

This PR is now up for feedback. The changes that were made over the last 2 years tried to keep the repository structure the same to make eventual reviewing easier. Whoever has concerns about the general idea of promoting this fork to main: please let me know, now is the time!

Full changelog with ~35 internal releases: https://github.com/prefrontalcortex/UnityGLTF/blob/master/UnityGLTF/Assets/UnityGLTF/CHANGELOG.md

The goal is to merge this PR in ~1-2 weeks.

This PR has been made possible by [prefrontal cortex](https://prefrontalcortex.de) and [Needle](https://twitter.com/NeedleTools).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/UnityGLTF?pullRequest=644) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/UnityGLTF?pullRequest=644) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: hybridherbst<br/>:x: marwie<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/UnityGLTF?pullRequest=644) it.</sub>

hybridherbst commented:
Added a couple more commits that have fixed issues people brought up (some of them old, some of them introduced by the PR).
I'd give it another few days for public testing and then merge back to master and properly set up OpenUPM releases :)

Ayfel commented:
Amazing, so get Fast is included here as well?

hybridherbst commented:
Big update that actually brings UnityGLTF back to being state-of-the-art in some areas!

- experimental support for KHR_animation_pointer export
- upon popular request, experimental URP (2020.3+) & BiRP (2021.2+) Shader Graph for export and import, `UnityGLTF/PBRGraph`
- approximated support for exporting and importing KHR_materials_transmission, KHR_materials_volume, KHR_materials_ior, KHR_materials_iridescence
- renderer features and post effect for rough refraction / transmission (for URP and BiRP)

Additionally, some more fixes went in for camera/lights export and import, material export heuristics, and helpful things like ProfilerMarkers.

keveleigh commented on UnityGLTF/Assets/UnityGLTF/Tests/Assets/SkyboxMaterial.mat:
```diff
@@ -76,7 +76,7 @@ Material:
         m_Offset: {x: 0, y: 0}
     - _Tex:
-        m_Texture: {fileID: 8900000, guid: 89f974ac697f8c348897a2cc7e13a754, type: 2}
+        m_Texture: {fileID: 0}
```
Lost a texture reference here. Intentional?

hybridherbst replied:
That was an old test asset, now removed

keveleigh commented on UnityGLTF/Assets/UnityGLTF/Runtime/Plugins/GLTFSerialization/Schema/MeshPrimitive.cs:
```diff
@@ -321,10 +319,12 @@ public override void Serialize(JsonWriter writer)
+		public const string WEIGHT = "WEIGHT";
 		public const string TANGENT = "TANGENT";
-		public const string INDICES = "INDICES";
+		public const string INDICES = "INDICIES";
```
This seems odd. The spec itself spells it `indices`: https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#indices-and-names

Why's this being re-spelled?

hybridherbst replied:
Fixed a while ago, thanks!

keveleigh commented on UnityGLTF/Assets/UnityGLTF/Runtime/Plugins/GLTFSerialization/Schema/Skin.cs:
```diff
@@ -30,6 +30,7 @@ public class Skin : GLTFChildOfRootProperty
 
 		public Skin()
 		{
+			Joints = new List<NodeId>();
```
I wonder if this might have unintended consequences due to assumptions in this class and others.
This class itself specifically checks `Joints != null` on line 50, though it checks `Joints != null && Joints.Count > 0` on line 113. Maybe line 50 just needs to be updated to also check `.Count` (and potentially anywhere else in the codebase that accesses `Joints`)?

keveleigh commented on UnityGLTF/Assets/UnityGLTF/Tests/UnityGLTFTests.asmdef:
```diff
@@ -1,12 +1,32 @@
+    ],
+    "versionDefines": [
+        {
+            "name": "Select...",
```
Mostly for my reference, but I'm unfamiliar with this `"Select..."` syntax! It normally points to a package name. What does this resolve to?

hybridherbst replied:
It just means that a local package reference is missing and was never properly set up in the first place. Just kept for legacy reasons

robertlong committed:
Add ability to export files and buffer views

Update to use streams

tram98 commented:
Wow! That is indeed a lot of great work! I just starred this repo a few days ago because I plan to use the GLTF importer in my project.
Thank you very much for that contribution @hybridherbst

hybridherbst committed:
add MSFT_lods export extension (wip)

Commit:
added stream size check to prevent endless loop + fixed bone joins import when using quantize

Commit:
added check when meshopt is missing but required

Commit:
code style cleanup

Commit:
removed space in const string

Commit:
added null checks for textures

Commit:
clean up and renaming

hybridherbst committed:
rough refraction 2023 fix

commit 89f0a9f09d6981c313a6547bd3a26db6407cbcd0
Author: pfcDorn <>
Date:   Fri Aug 4 12:14:59 2023 +0200

    removed space in const string

commit f092e7eaf927044320db7324037797072454be86
Author: pfcDorn <>
Date:   Fri Aug 4 12:06:34 2023 +0200

    code style cleanup

commit 565bbf9491f069423bc04fc4bdfc55f860c736cb
Author: pfcDorn <>
Date:   Fri Aug 4 08:45:35 2023 +0200

    prevent render pass execution when it's a preview camera

commit a326f7456031185d34dc965fcf6be7ea87a04300
Author: pfcDorn <dorn@prefrontalcortex.de>
Date:   Thu Aug 3 15:57:07 2023 +0200

    removed Downsampling option for unity 2023.1+

commit e9a050af6239a15234b4d155638c1a5fc9a633a2
Author: pfcDorn <dorn@prefrontalcortex.de>
Date:   Thu Aug 3 15:49:11 2023 +0200

    first draft - bilinear not working

hybridherbst committed:
fix ifdef

hybridherbst committed:
fix converter fallback when shader doesn't have glossiness

added null checks for textures

Commit:
fixed too early closed stream

hybridherbst committed:
Merge branch 'feature/mesh-opt-import' into version/2.x

hybridherbst committed:
Merge remote-tracking branch 'origin/Material-creation-fixes' into version/2.x

# Conflicts:
#	UnityGLTF/Assets/UnityGLTF/Runtime/Scripts/RenderPipelines/RoughRefractionFeature.cs

Commit:
added asset path check if it's empty

Commit:
ensure that textures always have a name

Commit:
added missing clear coat texture import

Commit:
added recalculatetangents when tangents are missing in mesh

Commit:
added missing specGlossinessTexCoord to importmaterials

Commit:
changed ExtTextureTransform.texCoord to nullable type. In some cases the texCoord is set in texture info and not in ExtTextureTransform

Commit:
added import settings to runtime importer (Normals, Tangents, SwapUVs)

Commit:
fixed interpolation animation import (InterpolationTest)

Commit:
code cleanup, some reverts from last commit, refactoring

Commit:
fixed wrong placed #endif

Commit:
export: ignore submeshes with 0 vertices

Commit:
fixes compiler warnings in 2023+

Commit:
code clean up

Commit:
code formating

Commit:
fixed not deserialized ExtTextureTransform inside material extensions

Commit:
added support for clearcoat roughness map and normal map (not supported in pbrgraph)

Commit:
make sure an emty texture name will get an empty name on export

Commit:
added "Texture_[index]" plus magic suffix with \0 to empty texture names on import

Commit:
added null check

Commit:
added uv channel export

robertdorn83 committed:
fixed typo

robertdorn83 committed:
fixed missing extexturetransform on textures within extensions

robertdorn83 committed:
added error log when trying to load an EXR texture

robertdorn83 committed:
import unreferenced materials and textures

Commit:
added null checks for importing non referenced textures

Commit:
revert broken changes

Commit:
add gltf images to import dependencies (was conflicting with ktx importer)

Commit:
updated changelog

Commit:
changelog formating

Commit:
fixed default material import

hybridherbst committed:
correct version check for KTX dependency

hybridherbst committed:
fix compile error in Stream.Write on 2020.3.43f1

hybridherbst committed:
fix: when sampling humanoid animation with KHR_animation_pointer enabled, additional tracks with MotionT, MotionQ and all muscle animations were incorrectly created

Commit:
added custom path combine method for relative paths

Commit:
fixes in case the same texture is used in many images

hybridherbst committed:
Merge branch 'debug/export-blendshapes-with-sampled-clips' into general-fixes

# Conflicts:
#	UnityGLTF/Assets/UnityGLTF/CHANGELOG.md
#	UnityGLTF/Assets/UnityGLTF/Runtime/Scripts/SceneExporter/ExporterAnimation.cs

hybridherbst committed:
add LICENSE

hybridherbst committed:
changelog/version/meta

Commit:
added output which gtlf file, in case a texture can't be imported

Commit:
remove smoothness input from shader

Commit:
fixed wrong file extension when extracting textures (Importer)

robertdorn83 committed:
added default gltf sampling to new loaded textures

Commit:
updated changelog

hybridherbst committed:
warn when importing legacy material KHR_materials_pbrSpecularGlossiness

hybridherbst committed:
log gather dependencies exceptions and use correct GLTFParser API

hybridherbst committed:
assign default values for texture only when a texture actually exists

hybridherbst committed:
unify and improve extension warning messages

hybridherbst committed:
remove obsolete constructors

hybridherbst committed:
cleanup

hybridherbst committed:
make sure _gltfFileName is set in an import context in the editor when loading from a stream

hybridherbst committed:
changelog and version

hybridherbst committed:
move UnityGltf folder to toplevel directory

hybridherbst committed:
remove legacy folders and test projects

hybridherbst committed:
remove legacy CI files that haven't been working in a long time

hybridherbst committed:
remove legacy files

hybridherbst committed:
license

hybridherbst committed:
clean up .npmignore for package releases

hybridherbst committed:
moved legacy tests to hidden legacy~ folder

hybridherbst committed:
rm legacy sample scenes that were used for old tests, not for samples

hybridherbst committed:
adjust changelog and version

hybridherbst committed:
adjust readme

hybridherbst committed:
fix: re-enable check for ShouldExportTransform with the exception of disabled prefab roots

hybridherbst committed:
changelog/version

hybridherbst committed:
add material creation menu and note about glTF editing being experimental

hybridherbst committed:
Hide experimental texture remapping by default

hybridherbst committed:
handle emissive color conversions in gamma space and with specific Unity shaders better

hybridherbst committed:
fix: improvements to light color and value export

hybridherbst committed:
fix light import color space

hybridherbst committed:
sanitize alpha cutoff value before writing

hybridherbst committed:
fix spot light import intensity

hybridherbst committed:
revert change: fix point light intensities for three and babylon

Gestaltor disagrees, report #429

hybridherbst committed:
feat: allow choice between implicit and explicit queue and surface type control

hybridherbst committed:
experiment: better sampling for rough refraction

hybridherbst committed:
convert light and material color spaces when exporting for animation

hybridherbst committed:
GltfRecorder color spaces

hybridherbst committed:
changelog/version

andybak commented:
What's the status of this PR. I can see lots of commits to the dev branch by @hybridherbst - does that work duplicate this PR? Should the PR be closed now?

hybridherbst commented:
Status is: we're mirroring changes in this PR and the dev branch. The last week have seen another significant push towards compatibility, readme, and documentation, and the plan is to merge the PR next week – finally.

hybridherbst committed:
compress textures on import to platform-specific default format

hybridherbst committed:
feat: per-material opt-in for glTF material editing

hybridherbst committed:
fix: indent level in the wrong place

hybridherbst committed:
add texture compression option

hybridherbst committed:
fix 2020.3 compilation error

marwie committed:
Prevent ArgumentException when animating same property name on multiple components on the same gameObject

hybridherbst committed:
bring instancing option back in PBRGraph

hybridherbst committed:
add tooltip, hide compression options when no textures are present

hybridherbst committed:
comments, changelog, version

Commit:
added import options: import BlendShape Names

Commit:
changed the OnBeforeAddAnimationData callback for a non destructive method to modify the animation data on saving

marwie committed:
change: always export root level objects (even if disabled or set to EditorOnly)

marwie committed:
version

robertdorn83 committed:
create default GLTFImportContext in runtime

robertdorn83 committed:
added ifdef for animationClip.SetCurve at runtime (only works with legacy animations clips)

Commit:
fixing importing warning of duplicated names

Commit:
fixed missing "blendshape." in animation import

Commit:
added UnityWebRequestLoader as default dataloader

Commit:
refactoring dataloader verify in sceneimporter

Commit:
removed unwanted #if Unity_editor

hybridherbst committed:
Merge remote-tracking branch 'pfc/blendshape-name-import-option' into dev

hybridherbst committed:
expose ExportAccessor(byte[]) and configurable overload for easier arbitrary data export

hybridherbst committed:
fix: meshes without submeshes and multiple materials behave the same as in Unity now (all materials are exported and result in multiple drawcalls)

hybridherbst committed:
add callback to GLTFRecorder with calculated translation bounds

commit d6f58fc75bfb24acba24924098af6bab6509269b
Author: pfcDorn <>
Date:   Mon Oct 16 09:33:32 2023 +0200

    removed animation bound marker in gltfrecorder, added postexport callback

commit 1e80426e1c60bbae74405c0ceed49a8a7bc364cc
Author: pfcDorn <>
Date:   Wed Sep 27 11:46:33 2023 +0200

    add callback for animation data editing, remove trim function

commit f846fe384cad72ff2e7d87d8851e08653c813c10
Author: pfcDorn <>
Date:   Wed Sep 20 13:40:32 2023 +0200

    changed to support multible time saving a recording

commit 2a27665b777ba10c92303a5fb1933ebcf4989202
Author: pfcDorn <>
Date:   Wed Sep 20 12:04:19 2023 +0200

    changed recording animation cutting

commit 91b8b50e648aec7e1b894235eff61ee148c4f81a
Author: pfcDorn <>
Date:   Tue Sep 19 15:32:30 2023 +0200

    Added recording cut functionality

hybridherbst committed:
fix: calling ExportAnimationClip with a humanoid clip but no Animator/Avatar resulted in an exception

properly warns now and returns a null clip

Futuremappermydud commented:
Any upodates?

hybridherbst commented:
Yup, we identified a few runtime import/export issues but we're mostly done.

hybridherbst committed:
changelog/version

hybridherbst committed:
fix importer context root not being set

Commit:
fixed wrong flipped triangles when it's required to generate them ("Fox"-Test model) + fixed wrong imported vertex data on submeshes

Commit:
removed old commented legacy code

Commit:
fixed draco ifdef compiler error + nullref fix

Commit:
reuse joints and weights on import for submeshes

Commit:
import normals when tangents are required

hybridherbst committed:
expose NodeCache

hybridherbst committed:
expose NodeCache

hybridherbst committed:
expose mesh cache and GetMeshId

hybridherbst committed:
Merge branch 'import-export-fixes' into dev

marwie committed:
Prevent animationclip being exported if referenced on animator + script with animatorcontroller reference (same clip added for the same object again from a script)

marwie committed:
Add support for exr texture export in correct format

previously it was exported with exr extension but actually encoded as PNG regardless

hybridherbst committed:
minor rename

hybridherbst committed:
move exr const to better place

hybridherbst committed:
fix humanoid importer inspector not being shown when model doesn't have animation data

fixes https://github.com/prefrontalcortex/UnityGLTF/issues/120

hybridherbst committed:
Merge branch 'prevent-multiple-animationclip-export' into dev

hybridherbst committed:
Merge branch 'gltf-exr-image-export' into dev

hybridherbst committed:
cleanup: move identical clip+node animation check into GetOrCreateAnimation

hybridherbst committed:
use the cache that also respects clip speed for GetAnimationId

marwie committed:
version

marwie committed:
export background color animation in KHR ptr

marwie committed:
camera backgroundcolor animation target should be the component

hybridherbst committed:
fix: curve sorting has to happen before checking r,g,b,a existance

hybridherbst committed:
fix: animated colors on arbitrary components, and camera background color, should be linear

marwie committed:
version

Commit:
added tangent.w value validation on export  > Meta Avatar tangents causes validation error. fixes #121

Commit:
fixed wrong settings assignment > recording anim. pointer was used for record in world space

Commit:
custom EnsureQuaterionContinuity for linear interpolation

Commit:
added missing line

hybridherbst committed:
fix: recalculate mesh bounds after applying import scale

hybridherbst committed:
Merge remote-tracking branch 'origin/timeline-fix' into dev

hybridherbst committed:
feat: allow enabling GPU instancing directly from the material importer

hybridherbst committed:
fix: sanitize animator state name on import

hybridherbst committed:
Merge branch 'feature/export-files' into dev

hybridherbst committed:
changelog/version

Commit:
import skinnedmesh scale fix

Commit:
import animation position scale fix

hybridherbst committed:
wip showing extensions

hybridherbst committed:
start switching to export plugin callbacks

hybridherbst committed:
add description, move plugin toggle to the left

hybridherbst committed:
renamings

hybridherbst committed:
refactor MSFT_lods export to use plugin structure

hybridherbst committed:
add plugins to show in UI with supported import/export features

hybridherbst committed:
settings cleanup

hybridherbst committed:
add warning icon and text when plugin can't function for some reason (e.g. missing package dependency)

hybridherbst committed:
cleanup

hybridherbst committed:
legacy API

hybridherbst committed:
move more internal settings to plugin/plugincontext

hybridherbst committed:
move animationpointer to plugin context

hybridherbst committed:
bring "UseAnimationPointer" API back with Obsolete attribute

hybridherbst committed:
better place to initialize plugins on import

hybridherbst committed:
add EnabledByDefault for plugins

hybridherbst committed:
move files

hybridherbst committed:
more renames

hybridherbst committed:
mark pbrSpecGloss as legacy

hybridherbst committed:
text clarification

hybridherbst committed:
fix using

hybridherbst committed:
feat: add BeforeNodeExport callback that allows for hierarchy and mesh modifications

hybridherbst committed:
give settings instances names and add them to the settings asset if possible

hybridherbst committed:
make sure to use default import settings for GLTFImporter instead of global settings

Need to be able to override them from the importer itself per asset.

hybridherbst committed:
move settings inspector to editor folder

hybridherbst committed:
Merge branch 'feature/configurable-extensions' of https://github.com/prefrontalcortex/UnityGLTF into feature/configurable-extensions

hybridherbst committed:
cleanup

hybridherbst committed:
apply importer settings overrides - no per-plugin settings yet

hybridherbst committed:
Merge branch 'feature/configurable-extensions' of https://github.com/prefrontalcortex/UnityGLTF into feature/configurable-extensions

hybridherbst committed:
rm old packages.config

hybridherbst committed:
move bake plugins into core

hybridherbst committed:
ensure plugin registering happening

hybridherbst committed:
show assembly name as header for plugins, allow pinging/opening plugin files directly

hybridherbst committed:
Merge branch 'feature/configurable-extensions' of https://github.com/prefrontalcortex/UnityGLTF into feature/configurable-extensions

hybridherbst committed:
move comments

hybridherbst committed:
changelog/version

hybridherbst committed:
Merge remote-tracking branch 'origin/import-scale-fix' into dev

hybridherbst committed:
Merge branch 'feature/configurable-extensions' into dev

hybridherbst committed:
rename import/export plugins to match UnityGLTF naming (GLTF instead of Gltf)

hybridherbst committed:
readme adjustments

hybridherbst committed:
code upgraders

hybridherbst committed:
move import scale factor into the import context

hybridherbst committed:
fix settings creation at runtime, fix compiler warnings at build time

hybridherbst committed:
sanitize animator state name

hybridherbst committed:
changelog/readme

hybridherbst committed:
add KHR_audio sample

hybridherbst committed:
fix: default property deserializer was missing for nested extras objects in MeshPrimitive

fixes https://github.com/prefrontalcortex/UnityGLTF/issues/126

hybridherbst committed:
fix ImportContext nullref

hybridherbst committed:
fix missing extras property deserialization in MeshPrimitive deserializer

fixes https://github.com/prefrontalcortex/UnityGLTF/issues/126 properly

hybridherbst committed:
use similar implementation pattern than GLTFMesh for clarity (avoids LINQ)

hybridherbst committed:
move blendshape target names to mesh, according to https://github.com/KhronosGroup/glTF/pull/1631

hybridherbst committed:
export menu cleanup, add selection export options to GameObject menu as well

hybridherbst committed:
changelog/version

Commit:
cache active colorspace because it can only be read from the main thread

Commit:
added missing using

Commit:
added IMultiThreadingSupported interface for IDataLoaders

robertdorn83 committed:
fixed bone weights issue with multiple sub meshes when using draco > bug in combine meshes

robertdorn83 committed:
removed Task.Run > to ensure the DataLoader (UnityWebRequest) is running on the mainthread

robertdorn83 committed:
Revert "added IMultiThreadingSupported interface for IDataLoaders"

This reverts commit 96a8b5765d496ffcec5e53d7486aaeb7c7cbf6f2.

robertdorn83 committed:
added asm define HAVE_GLTFAST for new com.unity.cloud.glfast

robertdorn83 committed:
fix when no tmpro is installed

robertdorn83 committed:
added asm define HAVE_TMPRO for com.unity.ugui v 2.0.0 which includes tmpro

Commit:
for editor import: convert normal maps to dxt5nm when its enabled in project settings

Commit:
added IsNormal to TextureCacheData

Commit:
fixed wrong placed variable declaration

Commit:
Normal encoding dependency

Commit:
Corrent normal encoding for runtime imported normals maps

Commit:
fixed normal map encoding conversion on asset import > ensure a RGBA format

hybridherbst committed:
add unity registry defines

hybridherbst committed:
fix: GLTFRecorder should respect existing settings for animation pointer usage

hybridherbst committed:
fix wrong string for RegisterCustomDependency and add AssetPostProcessor to cover more cases where a reimport might be necessary

hybridherbst committed:
changelog and version

hybridherbst committed:
looks like custom dependencies can only be set in an assetPostProcessor on 2021.3+

Commit:
ignore normal map conversion on export for runtime imported normal maps

Commit:
added missing material.getfloat("_RuntimeCreatedNormalMap" == 1)

Commit:
refactored ifdef

Commit:
renamed RuntimeCreatedNormalMap to NormalMapFormatXYZ in shader

Commit:
added unity version ifdef for hasFloat etc

hybridherbst committed:
better inspector to show mode overrides, rename _AutoSurfaceMode to clearer _OverrideSurfaceMode

hybridherbst committed:
fix variable name inside ifdef

hybridherbst committed:
changelog

hybridherbst committed:
move texture warnings GUI above tabbed inspector, log warning after editor import

hybridherbst committed:
changelog/version

Commit:
fixes correct tangent import for draco meshes

Commit:
fixes for non-readable normal textures when converting to dxt5nm format

robertdorn83 committed:
support for draco v5

robertdorn83 committed:
fixed linear color space when loading ktx textures

robertdorn83 committed:
draco ifdef fixes

robertdorn83 committed:
bone import fix

robertdorn83 committed:
set normalmapencodingconverter to internal + set texture name to replaced texture

robertdorn83 committed:
add asm define draco v5

hybridherbst committed:
fix tiling and offset properties being displayed despite _TEXTURE_TRANSFORM being off

hybridherbst committed:
fix nullref in texture fix checks when texture is null (e.g. specific name can't be found)

hybridherbst committed:
changelog/version

hybridherbst committed:
change plugin package note to info instead of warning to avoid confusion

hybridherbst committed:
rename GLTFPluginEditor and add package installation UI for optional compression packages

hybridherbst committed:
reword some of the package messaging

hybridherbst committed:
changelog

StephenHodgson commented on GLTFSerialization/External/glTF-pbrSpecularGlossiness/Lantern.bin:
Why did all of these get removed? Where did they go?

hybridherbst replied:
They have been removed for now because they were  
a) in the wrong place (test assets outside the Tests folder)  
b) tests didn't work even when I started working on the fork.  

Some recent changes to speed up certain models by 10-100x also resulted in the GLTFSerialization assembly having a dependency on Unity, which I'm not particularly happy with – but the performance improvements were worth it I think.

We are planning to bring better test capabilities back though :)  
![image](https://github.com/KhronosGroup/UnityGLTF/assets/2693840/fe74ae59-acbd-46c4-9397-7aba6338bbce)

StephenHodgson replied:
So no replacement?

StephenHodgson replied:
>  Khronos Sample Assets

Where are these located?

hybridherbst replied:
https://github.com/khronosGroup/gltf-sample-assets
